### PR TITLE
Register models for admin

### PR DESCRIPTION
**Pull request by thomasdenecker**
- [x] Register data models for admin
- [x] Register core models for admin
- [x] Customize `data` models listing / edition
- [x] Customize `core` models listing / edition
- [x] Add tests for all `ModelAdmin`
- [x] Rebase to `master`

### CODE REVIEW

File: apps/core/admin.py [hunk 1]
```diff
@@ -1 +1,15 @@
-# Register your models here.
+from django.contrib import admin
+from .models import Species, Strain, OmicsUnitType, \
+                    OmicsUnit, Pixel, Tag, Experiment, \
+                    Analysis, OmicsArea, Pixeler
```
Comment (jmaupetit) [inline apps/core/admin.py]: General remark about imports: i. I prefer using the tuple syntax, ii. with imports sorted alphabetically:

```python
from .models import (
    Analysis, Experiment, OmicsArea, OmicsUnit, OmicsUnitType, Pixel, Pixeler, 
    Species, Strain, Tag
)
```

But, in this case, if you will use all models, why not importing the `models` module itself?

```python
from . import models

admin.site.register(models.Analysis)
```

File: apps/data/admin.py [hunk 2]
```diff
@@ -1 +1,5 @@
-# Register your models here.
+from django.contrib import admin
+from .models import Repository, Entry
```
Comment (jmaupetit) [inline apps/data/admin.py]: If you agree with my first remark, apply the same idiom here.

**Comment by jmaupetit** [comment]
I think you can pursue your work in this branch/PR (with the **WIP** flag). Also, don't forget to rebase your branch:

```bash
# From the Data_ModelAdmin branch
$ git checkout master
$ git pull origin master
$ git checkout -
$ git rebase master
```

File: apps/core/admin.py (hunk 1 above)
Comment (thomasdenecker) [reply apps/core/admin.py]: The enumeration is part of the good practices in my remembers.

File: apps/data/admin.py (hunk 2 above)
Comment (thomasdenecker) [reply apps/data/admin.py]: Done

**Comment by thomasdenecker** [comment]
Normally, the rebase is ok.

File: apps/core/admin.py (hunk 1 above)
Comment (jmaupetit) [reply apps/core/admin.py]: Nope. Particularly when you will use a lot of module's content.

File: apps/core/models.py [hunk 3]
```diff
@@ -186,6 +186,12 @@ class Meta:
     def __str__(self):
         return str(self.reference)
```
Comment (jmaupetit) [inline apps/core/models.py]: I would prefer a composite str like (see comments below for more details about `UUIDModelMixin`):

```python
class OmicsUnit(UUIDModelMixin, models.Model):

    def __str__(self):
        return '{} ({}/{})'.format(
            self.get_short_uuid(),
            self.type,
            self.strain,
        )
```

File: apps/core/models.py [hunk 4]
```diff
@@ -398,6 +407,9 @@ class Meta:
     class MPTTMeta:
         order_insertion_by = ['name']
 
+    def __str__(self):
+        return str(self.name)
```
Comment (jmaupetit) [inline apps/core/models.py]: 👍

File: apps/core/admin.py [hunk 5]
```diff
@@ -1 +1,138 @@
...
+admin.site.register(models.Analysis, AnalysisAdmin)
+admin.site.register(models.Experiment, ExperimentAdmin)
+admin.site.register(models.OmicsArea, OmicsAreaAdmin)
+admin.site.register(models.OmicsUnit, OmicsUnitAdmin)
+admin.site.register(models.OmicsUnitType, OmicsUnitTypeAdmin)
+admin.site.register(models.Pixel, PixelAdmin)
+admin.site.register(models.Pixeler, PixelerAdmin)
+admin.site.register(models.Species, SpeciesAdmin)
+admin.site.register(models.Strain, StrainAdmin)
+admin.site.register(models.Tag, TagAdmin)
```
Comment (jmaupetit) [inline apps/core/admin.py]: For consistency and readability, please use the `@admin.register(MyModel)` decorator instead (like in `data/admin.py`).

File: apps/data/admin.py [hunk 6]
```diff
@@ -1 +1,25 @@
-# Register your models here.
+from django.contrib import admin
+from .models import Entry, Repository
+
+
+@admin.register(Repository)
+class RepositoryAdmin(admin.ModelAdmin):
+    search_fields = ['name']
+    ordering = ['name']
```
Comment (jmaupetit) [inline apps/data/admin.py]: I would prefer defining default order in the model, please declare another issue for this and we will address it later.

File: apps/data/admin.py [hunk 7]
```diff
@@ -1 +1,25 @@
...
+    search_fields = ['identifier', 'repository']
+    list_display = (
+        'identifier', 'description', 'url', 'repository_url'
+    )
+
+    def repository_url(self, obj):
+        return '<a href="%s" target="_blank">%s</a>' % \
+                (obj.repository.url, obj.repository.name)
+    repository_url.short_description = 'Repository'
+    repository_url.allow_tags = True
```
Comment (jmaupetit) [inline apps/data/admin.py]: I would remove this type of customization for now.

**Comment by jmaupetit** [comment]
Last question: why did you update the `docs/pixel-db.pdf` document?

**Comment by thomasdenecker** [comment]
It is not voluntary... Sorry :confused:

File: apps/core/admin.py [hunk 8]
```diff
@@ -1 +1,119 @@
...
+    )
+    list_filter = ('created_at', 'saved_at', 'tags', 'experiments__omics_area')
+
+
+@admin.register(models.Experiment)
+class ExperimentAdmin(admin.ModelAdmin):
+    list_display = (
+        'description', 'created_at', 'released_at', 'saved_at',
+    )
+
```
Comment (jmaupetit) [inline apps/core/admin.py]: Extra empty row here.

File: apps/core/admin.py [hunk 9]
```diff
@@ -1 +1,119 @@
...
+    def get_analysis(self, obj):
+        return obj.analysis.description
+    get_analysis.short_description = 'Analysis'
+
+
+@admin.register(models.Pixeler)
+class PixelerAdmin(admin.ModelAdmin):
+    list_display = ('username', 'first_name',
+                    'last_name', 'email', 'last_login'
+                    )
```
Comment (jmaupetit) [inline apps/core/admin.py]: Why changing indentation here?

```python
    list_display = (
        'username', 'first_name', 'last_name', 'email', 'last_login'
    )
```

File: apps/core/admin.py [hunk 10]
```diff
@@ -1 +1,119 @@
...
+@admin.register(models.OmicsUnit)
+class OmicsUnitAdmin(UUIDModelAdminMixin, admin.ModelAdmin):
+    list_display = (
+        'get_short_uuid', 'get_reference_identifier',
+        'strain', 'get_species', 'type', 'status'
+    )
+    list_filter = ['status', 'type', 'strain__species__name', ]
+
+    def get_species(self, obj):
+        return obj.strain.species.name
```
Comment (jmaupetit) [inline apps/core/admin.py]: `short_descriptiopn` is missing.

File: apps/core/models.py [hunk 11]
```diff
@@ -131,7 +131,28 @@ def __str__(self):
         return self.name
 
 
-class OmicsUnit(models.Model):
+class UUIDModelMixin(object):
+
+    def __str__(self):
+
+        print(self.__class__.__name__)
```
Comment (jmaupetit) [inline apps/core/models.py]: This print statement should be removed.

File: apps/core/models.py [hunk 12]
```diff
@@ -131,7 +131,28 @@ def __str__(self):
...
+        if self.__class__.__name__ == "OmicsUnit":
```
Comment (jmaupetit) [inline apps/core/models.py]: You should override the `__str__` method of the `OmicsUnit` model inheriting from the `UUIDModelMixin`. This is more explicit.

File: apps/data/admin.py [hunk 13]
```diff
@@ -1 +1,25 @@
...
+class EntryAdmin(admin.ModelAdmin):
+    search_fields = ['identifier', 'repository']
+    list_display = (
+        'identifier', 'description', 'url', 'repository_url'
+    )
+
+    def repository_url(self, obj):
+        return obj.repository.url
+    repository_url.short_description = 'Repository'
+    repository_url.allow_tags = True
```
Comment (jmaupetit) [inline apps/data/admin.py]: `repository_url` is not required. Really 😁

File: apps/core/models.py (hunk 11 above)
Comment (thomasdenecker) [reply apps/core/models.py]: Sorry... It was a test.

File: apps/core/factories.py [hunk 14]
```diff
@@ -49,3 +49,85 @@ class OmicsUnitFactory(DjangoModelFactory):
     class Meta:
         model = 'core.OmicsUnit'
         django_get_or_create = ('reference', 'strain')
+
+
+class PixelerFactory(DjangoModelFactory):
+
+    date_joined = Faker('date_object')
```
Comment (jmaupetit) [inline apps/core/factories.py]: It should be a `datetime_*` (see [Faker docs](https://faker.readthedocs.io/en/master/providers/faker.providers.date_time.html)). To check `AbstractUser` fields type, take a look at Django's source code:
https://github.com/django/django/blob/master/django/contrib/auth/models.py#L323

File: apps/core/factories.py [hunk 15]
```diff
@@ -49,3 +49,85 @@ class OmicsUnitFactory(DjangoModelFactory):
...
+    email = Faker('email')
+    first_name = Faker('first_name')
+    is_active = Faker('boolean')
+    is_staff = Faker('boolean')
+    is_superuser = Faker('boolean')
+    last_login = Faker('date_object')
```
Comment (jmaupetit) [inline apps/core/factories.py]: Same remark applies here.

File: apps/core/factories.py [hunk 16]
```diff
@@ -49,3 +49,85 @@ class OmicsUnitFactory(DjangoModelFactory):
...
+    name = Faker('word')
+    rght = Faker('pyint')
+    tree_id = Faker('pyint')
+
+    class Meta:
+        model = 'core.OmicsArea'
+        django_get_or_create = ('name',)
+
+
+class TagFactory(DjangoModelFactory):
```
Comment (jmaupetit) [inline apps/core/factories.py]: This whole factory will not work with the way you defined it: some fields are interdependent and your implementation is not consistent. Moreover, this factory is not required as you can define tags via a tagged model:

```python
analysis.tags = 'foo/bar/lol'
analysis.save()
```

File: apps/core/factories.py [hunk 17]
```diff
@@ -49,3 +49,85 @@ class OmicsUnitFactory(DjangoModelFactory):
...
+    slug = Faker('word')
+
+    class Meta:
+        model = 'core.Tag'
+        django_get_or_create = ('name', 'slug', 'path', 'label',)
+
+
+class ExperimentFactory(DjangoModelFactory):
+    omics_area = SubFactory(OmicsAreaFactory)
+    created_at = Faker('date_object')
```
Comment (jmaupetit) [inline apps/core/factories.py]: It should be a `datetime` instance.

File: apps/core/factories.py [hunk 18]
```diff
@@ -49,3 +49,85 @@ class OmicsUnitFactory(DjangoModelFactory):
...
+    description = Faker('text', max_nb_chars=300)
+    released_at = Faker('date_object')
+    saved_at = Faker('date_object')
```
Comment (jmaupetit) [inline apps/core/factories.py]: Same thing here.

File: apps/core/factories.py [hunk 19]
```diff
@@ -49,3 +49,85 @@ class OmicsUnitFactory(DjangoModelFactory):
...
+
+    class Meta:
+        model = 'core.Experiment'
+        django_get_or_create = ('omics_area', 'created_at')
+
+
+class AnalysisFactory(DjangoModelFactory):
+    pixeler = SubFactory(PixelerFactory)
+    created_at = Faker('date_object')
```
Comment (jmaupetit) [inline apps/core/factories.py]: date -> datetime.

File: apps/core/factories.py [hunk 20]
```diff
@@ -49,3 +49,85 @@ class OmicsUnitFactory(DjangoModelFactory):
...
+    description = Faker('text', max_nb_chars=300)
+    notebook = Faker('file_path', depth=1, category=None, extension=None)
+    saved_at = Faker('date_object')
```
Comment (jmaupetit) [inline apps/core/factories.py]: date -> datetime

File: apps/core/admin.py [hunk 21]
```diff
@@ -1 +1,120 @@
...
+    search_fields = ['name']
+    list_display = (
+        'name', 'description', 'get_species', 'entry_identifier'
+    )
+
+    def get_species(self, obj):
+        return (obj.species.name)
+    get_species.short_description = 'Species'
+
+    def entry_identifier(self, obj):
```
Comment (jmaupetit) [inline apps/core/admin.py]: Should be named: `get_entry_identifier`

File: apps/core/admin.py [hunk 22]
```diff
@@ -1 +1,120 @@
...
+    list_display = (
+        'name', 'description', 'reference', 'repository'
+    )
+
+
+@admin.register(models.Strain)
+class StrainAdmin(admin.ModelAdmin):
+    search_fields = ['name']
+    list_display = (
+        'name', 'description', 'get_species', 'entry_identifier'
```
Comment (jmaupetit) [inline apps/core/admin.py]: `s/entry_identifier/get_entry_identifier/`

File: apps/core/admin.py [hunk 23]
```diff
@@ -1 +1,120 @@
...
+
+    def get_species(self, obj):
+        return (obj.species.name)
+    get_species.short_description = 'Species'
+
+    def entry_identifier(self, obj):
+        if obj.reference is not None:
+            return (obj.reference.identifier)
+        else:
+            return ("-")
```
Comment (jmaupetit) [inline apps/core/admin.py]: To improve readability I would prefer:

```python
if obj.reference is None:
    return '-'

return obj.reference.identifier
```

File: apps/core/tests/test_admin.py [hunk 24]
```diff
@@ -0,0 +1,79 @@
...
+        self.Strain = factories.StrainFactory()
+        self.Strain_admin = admin.StrainAdmin(models.Strain, site)
+
+    def test_get_species(self):
+        self.assertEqual(
+            self.Strain_admin.get_species(self.Strain),
+            self.Strain.species.name
+        )
+
+    def test_entry_identifier_Not_None(self):
```
Comment (jmaupetit) [inline apps/core/tests/test_admin.py]: Should be named `test_get_entry_identifier_with_reference`

File: apps/core/tests/test_admin.py [hunk 25]
```diff
@@ -0,0 +1,79 @@
...
+        self.assertEqual(
+            self.Strain_admin.entry_identifier(self.Strain),
+            self.Strain.reference.identifier
+        )
+
+    def test_entry_identifier_None(self):
```
Comment (jmaupetit) [inline apps/core/tests/test_admin.py]: Should be named `test_get_entry_identifier_without_reference`

File: apps/core/admin.py [hunk 26]
```diff
@@ -1 +1,120 @@
...
+class PixelAdmin(admin.ModelAdmin):
+    list_display = (
+        'get_short_uuid', 'value', 'quality_score', 'omics_unit',
+        'get_analysis',
+    )
+    list_filter = (
+        'omics_unit__type', 'analysis__experiments__omics_area'
+    )
+
+    def get_analysis(self, obj):
```
Comment (jmaupetit) [inline apps/core/admin.py]: Should be named: `get_analysis_description`

File: apps/core/admin.py [hunk 27]
```diff
@@ -97,18 +97,18 @@ class SpeciesAdmin(admin.ModelAdmin):
...
         return (obj.species.name)
     get_species.short_description = 'Species'
 
-    def entry_identifier(self, obj):
-        if obj.reference is not None:
-            return (obj.reference.identifier)
+    def get_entry_identifier(self, obj):
+        if obj.reference is None:
+            return '-'
         else:
```
Comment (jmaupetit) [inline apps/core/admin.py]: This `else` statement is not required, as mentioned in my previous review.

File: apps/core/factories.py [hunk 28]
```diff
@@ -53,13 +53,13 @@ class Meta:
...
     email = Faker('email')
     first_name = Faker('first_name')
-    is_active = Faker('boolean')
-    is_staff = Faker('boolean')
-    is_superuser = Faker('boolean')
-    last_login = Faker('date_object')
+    is_active = Faker('pybool')
+    is_staff = Faker('pybool')
+    is_superuser = Faker('pybool')
+    last_login = Faker('date')
```
Comment (jmaupetit) [inline apps/core/factories.py]: This is not a `date` but a `datetime`.

File: apps/core/factories.py [hunk 29]
```diff
@@ -53,13 +53,13 @@ class Meta:
 
 class PixelerFactory(DjangoModelFactory):
 
-    date_joined = Faker('date_object')
+    date_joined = Faker('date')
```
Comment (jmaupetit) [inline apps/core/factories.py]: Like mentioned in my previous review: this is not a `date` but a `datetime`. See Faker documentation to choose the appropriate provider: https://faker.readthedocs.io/en/master/providers/faker.providers.date_time.html

Comment (jmaupetit) [reply apps/core/factories.py]: Take time to check model fields types so that you choose relevant Faker providers.

**Pull request merged**